updates cancelToken to AbortSignal

AlanCea68 opened the pull request:
Updates deprecated `cancelToken` used in the template to `AbortSignal` through conditional use of the `IS_AXIOS_VERSION_ZERO` variable. All comments and variables in the mustache template will be updated in accordance with the value provided by the user (note that `IS_AXIOS_VERSION_ZERO` is `0` by default/no value provided). 

In other words, the `IS_AXIOS_VERSION_ZERO` variable will not only dictate which version type of the request headers will be used:
 - old: `Record<string, unknown>`, 
 - new: `RawAxiosHeaders`,
 
but also the way canceled requests are handled: 
- old : `cancelToken`
- new: `AbortSignal`

https://axios-http.com/docs/cancellation
https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal

The `pascalCase` function is also updated to account for acronyms in the names provided. If a name contains an acronym, the acronym case persists in capitals and proceeds with pascal case as normal (ex. "TESTTest" -> "TESTTest")

AlanCea68 committed:
updates cancelToken to AbortController

YangJonathan commented on templates/service.mustache:
```diff
@@ -5,7 +5,7 @@
 /* eslint-disable  @typescript-eslint/no-unused-vars */
 import * as ApiModelTypes from "{{{TYPES_DIRECTORY}}}";
-import axios, { AxiosResponse, CancelToken{{^IS_AXIOS_VERSION_ZERO}}, RawAxiosRequestHeaders{{/IS_AXIOS_VERSION_ZERO}} } from "axios";
+import axios, { AxiosResponse, RawAxiosRequestHeaders{{/IS_AXIOS_VERSION_ZERO}} } from "axios";
```
So I'm observing an error when trying to generate the files. 
![image](https://github.com/AlanCea68/tag-api-codegen/assets/83241918/ea9f5edd-4bce-4cd1-91ed-9422327cc524)

---
So `{{/IS_AXIOS_VERSION_ZERO}}` is a mustache tag that defines the end of a "section" indicated by the slash `/`. Think of this like the closing tag of a JSX/TSX element "`</component>`".

A "section" in mustache is a block of template between an opening and a closing tag that is conditionally rendered with the conditional logic being based off the opening tag for which there are two types. 



- First, a normal section tag which starts with a pound symbol `{{#KEY}}`, which means if the key does not exist, or exists and has a value of null, undefined, false, 0, or NaN, or is an empty string or an empty list, the block will not be rendered.
- Second, an inverted section tag which starts with the carrot symbol `{{^KEY}}`, which means the block of an inverted section is rendered only if the value of that section's tag is null, undefined, false, falsy or an empty list. 

In the case before this modification the opening tag was `{{^IS_AXIOS_VERSION_ZERO}}` which has been removed, this is a violation of mustache rules and hence the error. You can learn more about how the mustache template works visiting this page: https://github.com/janl/mustache.js and scrolling down. 


What we need to figure out is if we need to keep the inverted opening tag or put in the normal opening tag. 

It looks to me like based off the changes then it would still be ok for us to keep the inverted tag as the abort controller is supported prior to v1 which is the value for which `IS_AXIOS_VERSION_ZERO` is based on. 

Tagging @kellirains for inputs here so whatever we do here aligns with her recent changes with supporting differing axios versions. Can you help review to make sure the changes here doesn't conflict with your recent logic?

kellirains replied:
Sure! My changes with the IS_AXIOS_VERSION_ZERO variable added a conditional type RawAxiosRequestHeaders when the user provides an Axios version and it is NOT zero, as it's unsupported in 0.x versions but versions > 1.0 give a type error when using `Record<string, unknown>` (the previous type for headers). You can see the conditional headers types being generated on line 57.

So we do need the entire block `{{^IS_AXIOS_VERSION_ZERO}} RawAxiosRequestHeaders{{/IS_AXIOS_VERSION_ZERO}}` and it needs to be inverted, which adds `, RawAxiosRequestHeaders` into the import statement if the version is NOT zero.

AlanCea68 replied:
This makes SO much sense now! Thanks team.

YangJonathan reviewed (changes requested):
Observing an issue, see comments. Glad to see more diving into taggem! It's a fun tool to traverse into.

YangJonathan commented:
Note: We'll want to run taggem with this new change to ensure that files are still successfully generated.

kellirains commented:
> Note: We'll want to run taggem with this new change to ensure that files are still successfully generated.

I'm attempting to test Alan's fork of the taggem tool and having some yarn issues, but @AlanCea68 if you want a demo of running it and generating api files lmk :D

YangJonathan commented:
> > Note: We'll want to run taggem with this new change to ensure that files are still successfully generated.
> 
> I'm attempting to test Alan's fork of the taggem tool and having some yarn issues, but @AlanCea68 if you want a demo of running it and generating api files lmk :D

try npm instead.

AlanCea68 committed:
updates mustache conditional tag, fixes readme typos

kellirains commented:
> > I'm attempting to test Alan's fork of the taggem tool and having some yarn issues, but @AlanCea68 if you want a demo of running it and generating api files lmk :D
> 
> try npm instead.

It was 100% user error

Builds fine now! Generates fine both for 0.x and 1.x templates. I am curious if we should grab an old version of alldataapp and test the cancelling with the abort signal just to make sure this "new" service file will work for other users using the taggem tool. I think we were on 0.22 before the Axios update

AlanCea68 committed:
adds JSDoc comment for optional axiosVersion param

kellirains commented:
One note - if you're using the newest api from server (https://api.dev.alldataapp.com/v1/docs/openapi) you may get an error - Cannot read properties of undefined (reading 'type'). I replaced the `pascalCase` function with 
```
const pascalCase = (string) => {
	return _.upperFirst(string);
	// return _.upperFirst(_.camelCase(string));
};
```

This is an issue I started to look into but never got a solution that I was happy with. It stems from pascalCase transforming all caps acronyms like VCRServiceStuff to VcrServiceStuff and then being unable to find the definition for VcrServiceStuff because it is VCRServiceStuff. This fixes it, but I don't feel like it's a good enough permanent solution to create a pr.

AlanCea68 commented:
> One note - if you're using the newest api from server (https://api.dev.alldataapp.com/v1/docs/openapi) you may get an error - Cannot read properties of undefined (reading 'type'). I replaced the `pascalCase` function with
> 
> ```
> const pascalCase = (string) => {
> 	return _.upperFirst(string);
> 	// return _.upperFirst(_.camelCase(string));
> };
> ```
> 
> This is an issue I started to look into but never got a solution that I was happy with. It stems from pascalCase transforming all caps acronyms like VCRServiceStuff to VcrServiceStuff and then being unable to find the definition for VcrServiceStuff because it is VCRServiceStuff. This fixes it, but I don't feel like it's a good enough permanent solution to create a pr.

What about this alternative? 
```
const pascalCase = (string) => {
    // First, split the string into words
    const words = string.split(/(?=[A-Z])|[\s-_]+/).filter(word => word.length > 0);
    
    return words.map((word, index) => {
        if (word.toUpperCase() === word && word.length > 1) {
            // If the word is all uppercase and more than one character, assume it's an acronym
            return word;
        } else if (index === 0 || word.length > 1) {
            // Capitalize the first letter, lowercase the rest, unless it's a single character
            return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
        } else {
            // For single characters (except the first word), just uppercase
            return word.toUpperCase();
        }
    }).join('');
};
```

kellirains commented:
> 
> What about this alternative?
> 
> ```
> const pascalCase = (string) => {
>     // First, split the string into words
>     const words = string.split(/(?=[A-Z])|[\s-_]+/).filter(word => word.length > 0);
>     
>     return words.map((word, index) => {
>         if (word.toUpperCase() === word && word.length > 1) {
>             // If the word is all uppercase and more than one character, assume it's an acronym
>             return word;
>         } else if (index === 0 || word.length > 1) {
>             // Capitalize the first letter, lowercase the rest, unless it's a single character
>             return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
>         } else {
>             // For single characters (except the first word), just uppercase
>             return word.toUpperCase();
>         }
>     }).join('');
> };
> ```

This works, can you break down your regex a little bit more? I'm having trouble understanding what [\s-_] matches. 

I'm not sure if the first check is necessary, as VCRIotRule is split into [ 'V', 'C', 'R', 'Iot', 'Rule' ] instead of ['VCR', 'Iot', 'Rule']. With your regex, would it ever split on an all caps word? Which is what your first case is handling, I think. 

I can create a branch that adds some logs to show the original problem if you want to try to tackle it. I think the original purpose of the pascal case function was to create consistently cased names, but in the future if server gives us some wild naming convention I think we should just handle it and keep it as they give it instead of transforming it at all, which is what I was playing around with.

AlanCea68 commented:
> > What about this alternative?
> > ```
> > const pascalCase = (string) => {
> >     // First, split the string into words
> >     const words = string.split(/(?=[A-Z])|[\s-_]+/).filter(word => word.length > 0);
> >     
> >     return words.map((word, index) => {
> >         if (word.toUpperCase() === word && word.length > 1) {
> >             // If the word is all uppercase and more than one character, assume it's an acronym
> >             return word;
> >         } else if (index === 0 || word.length > 1) {
> >             // Capitalize the first letter, lowercase the rest, unless it's a single character
> >             return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
> >         } else {
> >             // For single characters (except the first word), just uppercase
> >             return word.toUpperCase();
> >         }
> >     }).join('');
> > };
> > ```
> 
> This works, can you break down your regex a little bit more? I'm having trouble understanding what [\s-_] matches.
> 
> I'm not sure if the first check is necessary, as VCRIotRule is split into [ 'V', 'C', 'R', 'Iot', 'Rule' ] instead of ['VCR', 'Iot', 'Rule']. With your regex, would it ever split on an all caps word? Which is what your first case is handling, I think.
> 
> I can create a branch that adds some logs to show the original problem if you want to try to tackle it. I think the original purpose of the pascal case function was to create consistently cased names, but in the future if server gives us some wild naming convention I think we should just handle it and keep it as they give it instead of transforming it at all, which is what I was playing around with.

The `[\s-_]` checks for whitespaces (using the escape notation`\s`), dashes and underscores and splits at that. 
The lookahead notation is also used. It looks at the next character and keeps going until it changes, then splits. This is to account for intentional acronyms, like VCRService. This example would be split `[V, C, R, Service]`, then joins at the end, creating `VCRService`, which meets the requirements for pascal case. 

Since I don't anticipate other special characters or numbers being used for this application, I think this should be robust enough. Here, I'm only accounting for spaces, dashes and underscores to be converted to PascalCase. Acronyms should persist and be left as is. 

All caps words should also be supported here. The first conditional checks for that. 
![image](https://github.com/AlanCea68/tag-api-codegen/assets/113615525/867126e5-4ded-4789-940b-f9fbf285b514)


I can also leave as is and create a new ticket though. This task can definitely be postponed and planned in more detail at a later time.

AlanCea68 committed:
updates conditional cancelToken/AbortSignal params & response properties

AlanCea68 committed:
updates conditinoal comments and imports

AlanCea68 commented:
> > > I'm attempting to test Alan's fork of the taggem tool and having some yarn issues, but @AlanCea68 if you want a demo of running it and generating api files lmk :D
> > 
> > 
> > try npm instead.
> 
> It was 100% user error
> 
> Builds fine now! Generates fine both for 0.x and 1.x templates. I am curious if we should grab an old version of alldataapp and test the cancelling with the abort signal just to make sure this "new" service file will work for other users using the taggem tool. I think we were on 0.22 before the Axios update

Just tested with an old dev build branch from last year and it worked on my local. Thanks for this suggestion!

AlanCea68 committed:
fixes type typo

AlanCea68 committed:
updates typo

YangJonathan reviewed (approved):
Nice work Alan!

YangJonathan commented:
> > > What about this alternative?
> > > ```
> > > const pascalCase = (string) => {
> > >     // First, split the string into words
> > >     const words = string.split(/(?=[A-Z])|[\s-_]+/).filter(word => word.length > 0);
> > >     
> > >     return words.map((word, index) => {
> > >         if (word.toUpperCase() === word && word.length > 1) {
> > >             // If the word is all uppercase and more than one character, assume it's an acronym
> > >             return word;
> > >         } else if (index === 0 || word.length > 1) {
> > >             // Capitalize the first letter, lowercase the rest, unless it's a single character
> > >             return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
> > >         } else {
> > >             // For single characters (except the first word), just uppercase
> > >             return word.toUpperCase();
> > >         }
> > >     }).join('');
> > > };
> > > ```
> > 
> > 
> > This works, can you break down your regex a little bit more? I'm having trouble understanding what [\s-_] matches.
> > I'm not sure if the first check is necessary, as VCRIotRule is split into [ 'V', 'C', 'R', 'Iot', 'Rule' ] instead of ['VCR', 'Iot', 'Rule']. With your regex, would it ever split on an all caps word? Which is what your first case is handling, I think.
> > I can create a branch that adds some logs to show the original problem if you want to try to tackle it. I think the original purpose of the pascal case function was to create consistently cased names, but in the future if server gives us some wild naming convention I think we should just handle it and keep it as they give it instead of transforming it at all, which is what I was playing around with.
> 
> The `[\s-_]` checks for whitespaces (using the escape notation`\s`), dashes and underscores and splits at that. The lookahead notation is also used. It looks at the next character and keeps going until it changes, then splits. This is to account for intentional acronyms, like VCRService. This example would be split `[V, C, R, Service]`, then joins at the end, creating `VCRService`, which meets the requirements for pascal case.
> 
> Since I don't anticipate other special characters or numbers being used for this application, I think this should be robust enough. Here, I'm only accounting for spaces, dashes and underscores to be converted to PascalCase. Acronyms should persist and be left as is.
> 
> All caps words should also be supported here. The first conditional checks for that. ![image](https://private-user-images.githubusercontent.com/113615525/343922872-867126e5-4ded-4789-940b-f9fbf285b514.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA3MTQ5MTYsIm5iZiI6MTcyMDcxNDYxNiwicGF0aCI6Ii8xMTM2MTU1MjUvMzQzOTIyODcyLTg2NzEyNmU1LTRkZWQtNDc4OS05NDBiLWY5ZmJmMjg1YjUxNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcxMVQxNjE2NTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMDg2MTg2OWJlM2ZlYWUzOGE4NzVkOTlhNGFkYjNmODk0N2NkM2QxNGZiZDY5MDVjZDA4NjljYmU3Nzc0NDRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.woZOboJl1_e0tX_elR4YLAXfQb-6-308SZSzTM1NNsI)
> 
> I can also leave as is and create a new ticket though. This task can definitely be postponed and planned in more detail at a later time.


We'll want this as part of this PR so that taggem can handle capitalized acronyms when pascal-casing a name. You guys nailed the issue spot on.

AlanCea68 committed:
updates pascalCase function to account and have acronyms persist their case

AlanCea68 committed:
adds example comment

YangJonathan committed:
Resolve error with empty interfaces when parsing Array schema types

YangJonathan committed:
Resolve errors

YangJonathan committed:
Resolve runtime models not being generated

YangJonathan committed:
Remove console.log

AlanCea68 committed:
simplifies pascalCase logic

AlanCea68 committed:
refactors (DRYs up) apiModelTypes object property creation, refactors pascalCase function
